src/query/interpreter.cpp

```diff
@@ -1871,6 +1879,40 @@ PreparedQuery PrepareIsolationLevelQuery(ParsedQuery parsed_query, bool in_expli
       RWType::NONE};
 }
 
+PreparedQuery PrepareStorageModeQuery(ParsedQuery parsed_query, const bool in_explicit_transaction,
+                                      InterpreterContext *interpreter_context) {
+  if (in_explicit_transaction) {
+    throw StorageModeModificationInMulticommandTxException();
+  }
+
+  auto *storage_mode_query = utils::Downcast<StorageModeQuery>(parsed_query.query);
+  MG_ASSERT(storage_mode_query);
+
+  const auto storage_mode = ToStorageStorageMode(storage_mode_query->storage_mode_);
+
+  auto callback = [storage_mode, interpreter_context]() -> std::function<void()> {
+    auto exists_active_transaction = interpreter_context->interpreters.WithLock([](const auto &interpreters_) {
+      return std::any_of(interpreters_.begin(), interpreters_.end(), [](const auto &interpreter) {
+        return interpreter->transaction_status_.load() == TransactionStatus::ACTIVE;
+      });
+    });
+    if (exists_active_transaction) {
+      throw StorageModeModificationInMultiTxException();
+    }
+
+    return [interpreter_context, storage_mode] { interpreter_context->db->SetStorageMode(storage_mode); };
```

Comment:
Yes, you are right, I didn't think about that. Good catch. I will move this inside Lock part, so that we set state when we have lock. 

I think we should have implemented a transactional queue with locks and not atomics :joy:

Reply:
Also, it might happen that an already checked interpreter starts to execute another transaction, so this is only a best effort thing to do. The storage should acquire the main lock, so it would ensure that no transactions are active in the database.

Reply:
You are right here @antaljanosbenjamin added this. @as51340 check Benjo's comment and I added this check.